crates/masonry/src/text2/edit.rs

```diff
@@ -101,38 +101,49 @@ impl<T: EditableText> TextEditor<T> {
-                                ));
-                                Handled::Yes
-                            } else {
-                                debug_panic!("Got text input event whilst not focused");
-                                Handled::No
-                            }
+                            let selection = self.selection.selection.unwrap_or(Selection {
+                                anchor: 0,
+                                active: 0,
+                                active_affinity: Affinity::Downstream,
+                                h_pos: None,
+                            });
+                            // TODO: We know this is not the fullest model of copy-paste, and that we should work with the inner text
+                            // e.g. to put HTML code if supported by the rich text kind
+                            let c = ' ';
+                            self.text_mut().edit(selection.range(), c);
+                            self.selection.selection = Some(Selection::caret(
+                                selection.min() + c.len_utf8(),
+                                // We have just added this character, so we are "affined" with it
+                                Affinity::Downstream,
+                            ));
+                            let contents = self.text().as_str().to_string();
+                            ctx.submit_action(Action::TextChanged(contents));
+                            Handled::Yes
+                        }
+                        Key::Named(NamedKey::Enter) => {
+                            let contents = self.text().as_str().to_string();
+                            ctx.submit_action(Action::TextEntered(contents));
+                            Handled::Yes
                         }
                         Key::Named(_) => Handled::No,
                         Key::Character(c) => {
-                            if let Some(selection) = self.selection.selection {
-                                // TODO: We know this is not the fullest model of copy-paste, and that we should work with the inner text
-                                // e.g. to put HTML code if supported by the rich text kind
-                                self.text_mut().edit(selection.range(), &**c);
-                                self.selection.selection = Some(Selection::caret(
-                                    selection.min() + c.len(),
-                                    // We have just added this character, so we are "affined" with it
-                                    Affinity::Downstream,
-                                ));
-                                Handled::Yes
-                            } else {
-                                debug_panic!("Got text input event whilst not focused");
-                                Handled::No
-                            }
+                            let selection = self.selection.selection.unwrap_or(Selection {
```

Comment:
Ok, so this unwrap is handling the case where the text gets reset by the driver?

I don't think the caret at index 0 is the right choice there. If the text is being changed, I think that's a sufficient event to drop the focus; yes that's disruptive, but logically the Textbox with a different content is a new text field, even if it holds the same location in the layout.

Reply:
Perhaps. It still a pattern which we should try and discourage if it all possible.

Can we force the caller of `set_text` to provide the new caret position?

Ok, I can imagine this being right for a textbox like at the top of a spreadsheet? I guess you could also be moving down a carousel? Although logically in that case, you're moving focus to the next item; it just happens to have moved to the same place as the previously focused item

Reply:
You make a good point, but I don't really feel in the mindset to figure out the ideal DX for this. For now I think I'll just add a TODO comment.